### ISSUE TITLE
exception on CopyPixels ; buffer update

### ISSUE THREAD

**Issue by erutanrevol** [T1T4N/NVNC#1]
Hello! Your project is cool, I am watching it for some time.

This version has an exception in CopyPixel util, then you have 1440x900 monitor resolution.

And I don't know, maybe it is just me, but even in other resolution (1366x768) there is no exception, but framebuffer does not update. So client can see only 1st picture (and mouse cursor works on server, but client cant see it)

**Comment by tgiachi** [comment]
Hi!, I have same problem in dual monitor

**Comment by Sachin4dotnet** [comment]
Same problem faced by me also. Let me know if you found any work around..

**Comment by edernucci** [comment]
Same problem, but only when I'm connected from my mac (it's on a remote server). When I connect from my PC I have no problems.

**Comment by isupovs** [comment]
Yep. IndexOutOfRangeException

**Comment by oloffr** [comment]
Get rid of the IndexOutOfRangeException, but notably this is not the root cause.

PixelGrabber.cs:
        public static int[] CopyPixels(int[] pixels, int scanline, int x, int y, int w, int h)

Replace:
                ourPixels[i] = pixels[p];
with:
                if (p >= pixels.Length) {
                    p = pixels.Length - 1;
                }
                ourPixels[i] = pixels[p];

**Comment by dhilgarth** [comment]
The actual reason is the incorrect parameters for the two last calls to `CopyPixels` inside `QuadNode.GenerateChildren`. Fix it with this version:

```
public void GenerateChildren()
        {
            int westX = Bounds.X;   //Keep X and Y just in case, needed on client side
            int westY = Bounds.Y;
            int westWidth = Bounds.Width / 2;
            int westHeight = Bounds.Height / 2;

            int eastX = westX + westWidth;
            int eastY = westY;
            int eastWidth = Bounds.Width - westWidth;
            int eastHeight = Bounds.Height - westHeight;

            Rectangle2 nw = new Rectangle2(westX, westY, westWidth, westHeight);
            childrenData[(int)Direction.NW] = PixelGrabber.CopyPixels(NodeData, Bounds.Width, 0, 0, westWidth, westHeight);
            childrenRect[(int)Direction.NW] = nw;

            int parentWidth = Bounds.Width;
            int scanline = parentWidth;
            int size = westWidth * westHeight;
            int jump = scanline - westWidth;
            int s = 0;
            int p = 0 * scanline + 0;
            Trace.WriteLine("My offset: " + p);
            for (int i = 0; i < size; i++, s++, p++)
            {
                if (s == westWidth)
                {
                    s = 0;
                    p += jump;
                }
            }
            Trace.WriteLine("My end: " + --p);

            Rectangle2 ne = new Rectangle2(eastX, eastY, eastWidth, eastHeight);
            
            // TODO: Keep relative pixel start and end instead of copying the part of the array
            // The current implementation is a naive one and very slow, but it works as intended
            childrenData[(int)Direction.NE] = PixelGrabber.CopyPixels(NodeData, Bounds.Width, westWidth, 0, eastWidth, eastHeight);
            childrenRect[(int)Direction.NE] = ne;

            s = 0;
            p = 0 * scanline + westWidth;
            Trace.WriteLine("My offset: " + p);
            for (int i = 0; i < size; i++, s++, p++)
            {
                if (s == westWidth)
                {
                    s = 0;
                    p += jump;
                }
            }
            Trace.WriteLine("My end: " + --p);


            Rectangle2 sw = new Rectangle2(westX, westY + westHeight, westWidth, eastHeight);
            childrenData[(int)Direction.SW] = PixelGrabber.CopyPixels(NodeData, Bounds.Width, 0, westHeight, westWidth, eastHeight);
            childrenRect[(int)Direction.SW] = sw;

            Rectangle2 se = new Rectangle2(eastX, eastY + westHeight, eastWidth, eastHeight);
            childrenData[(int)Direction.SE] = PixelGrabber.CopyPixels(NodeData, Bounds.Width, westWidth, westHeight, eastWidth, eastHeight);
            childrenRect[(int)Direction.SE] = se;
        }
```

**Comment by ali-hm** [comment]
Is there any solution for this issue?

**Comment by T1T4N** [comment]
Fixed in [630c44f](https://github.com/T1T4N/NVNC/commit/630c44f8014bdd4cac771223e4d4f4fd968178bd).
